core/chains/evm/gas/models.go

```diff
@@ -261,32 +264,53 @@ func (e *evmFeeEstimator) L1Oracle() rollups.L1Oracle {
 	return e.EvmEstimator.L1Oracle()
 }
 
-func (e *evmFeeEstimator) GetFee(ctx context.Context, calldata []byte, feeLimit uint64, maxFeePrice *assets.Wei, opts ...feetypes.Opt) (fee EvmFee, chainSpecificFeeLimit uint64, err error) {
+func (e *evmFeeEstimator) GetFee(ctx context.Context, calldata []byte, feeLimit uint64, maxFeePrice *assets.Wei, toAddress *common.Address, opts ...feetypes.Opt) (fee EvmFee, chainSpecificFeeLimit uint64, err error) {
+	// Create call msg for gas limit estimation, if needed
+	callMsg := ethereum.CallMsg{
+		To:   toAddress,
+		Data: calldata,
+	}
+
 	// get dynamic fee
 	if e.EIP1559Enabled {
 		var dynamicFee DynamicFee
 		dynamicFee, err = e.EvmEstimator.GetDynamicFee(ctx, maxFeePrice)
 		if err != nil {
 			return
 		}
-		chainSpecificFeeLimit, err = commonfee.ApplyMultiplier(feeLimit, e.geCfg.LimitMultiplier())
 		fee.DynamicFeeCap = dynamicFee.FeeCap
 		fee.DynamicTipCap = dynamicFee.TipCap
-		return
+		chainSpecificFeeLimit = feeLimit
+		// Set call msg fields with dynamic fee fields for gas limit estimation, if needed
+		callMsg.GasFeeCap = fee.DynamicFeeCap.ToInt()
+		callMsg.GasTipCap = fee.DynamicTipCap.ToInt()
+	} else {
+		// get legacy fee
+		fee.Legacy, chainSpecificFeeLimit, err = e.EvmEstimator.GetLegacyGas(ctx, calldata, feeLimit, maxFeePrice, opts...)
+		if err != nil {
+			return
+		}
+		// Set call msg fields with legacy fee field for gas limit estimation, if needed
+		callMsg.GasPrice = fee.Legacy.ToInt()
 	}
 
-	// get legacy fee
-	fee.Legacy, chainSpecificFeeLimit, err = e.EvmEstimator.GetLegacyGas(ctx, calldata, feeLimit, maxFeePrice, opts...)
-	if err != nil {
-		return
+	if e.geCfg.EstimateGasLimit() {
+		callMsg.Gas = chainSpecificFeeLimit
+		estimatedGasLimit, estimateErr := e.ethClient.EstimateGas(ctx, callMsg)
+		if estimateErr != nil {
+			// Do not return error if estimating gas failed, we can still use the provided limit instead since it is an upper limit
+			e.lggr.Errorw("failed to estimate gas limit. falling back to provided gas limit.", "toAddress", toAddress, "data", calldata, "gasLimit", chainSpecificFeeLimit, "error", estimateErr)
+		} else {
+			chainSpecificFeeLimit = estimatedGasLimit
```

Comment:
This estimatedGasLimit is the estimated Gas returned to us by the chain.
Due to fluctuations or errors, the actual gas used may be slightly more than this, causing the actual Tx to fail.
I think we should add a buffer on top of what the RPC returns to us.
Make that buffer a config value, EstimatedGasBuffer, set to default of 25%.
This protects us from fluctuations.

Also, you need to ensure that feeLimit > estimatedGasLimit. If not, then we can error out. Basically the chain is saying the gas used will be more than what the callers allows us to use, thus making this Tx infeasible.

Lastly, ensure that SuggestedGasLimit = MIN(feeLimit, estimatedGasLimit+Buffer)

Reply:
I wanted to avoid adding another config so I was using the existing `LimitMultiplier` as the buffer. I could see the distinction between the 2 causing confusion because `LimitMultiplier` was expected to be added on top of the provided limit but `estimated gas * EstimatedGasBuffer` would be expected to stay under the provided limit